### ISSUE TITLE
PackageKit not automatically refreshing updates

### ISSUE THREAD

**Issue by christiantosta** [PackageKit/PackageKit#588]
I'm running Fedora 37 with KDE Discover, which uses the PackageKit infrastructure to get updates, and I came across the following problem: if I run a "dnf upgrade" the updates are shown normally, however they are not shown in Discover.

A bug was opened with KDE about this and it became evident that the problem is related to PackageKit or some plugin related to the distribution.

https://bugs.kde.org/show_bug.cgi?id=461962
![discover-or-packagekit-bug](https://user-images.githubusercontent.com/7252968/204707377-dae34487-2908-42d3-83d6-ac5865448988.png)

So, as sugested by Nate from KDE, I ran some pkcon commands (see attachments).

As you can see, it looks like PackageKit is not automatically updating the update lists.
[pkcon-commands-output.txt](https://github.com/PackageKit/PackageKit/files/10119463/pkcon-commands-output.txt)

The packagekitd is running normally as root user.

**Comment by ximion** [comment]
This is working perfectly fine on Debian/Ubuntu, so I wonder whether this is actually an issue with the DNF backend in PK - maybe DNF doesn't notify the daemon that there were changes to the cache?

**Comment by christiantosta** [comment]
Please see #9 and Fedora bugzilla#1950041. Maybe is this related to that cache-age issues?

https://bugzilla.redhat.com/show_bug.cgi?id=1950041

PS: If I run pkcon with -c option, the cache refreshes and updates turns available to Discover.

**Comment by ximion** [comment]
> PS: If I run pkcon with -c option, the cache refreshes and updates turns available to Discover.

That's quite telling! This needs someone with DNF experience to resolve then...

**Comment by aleasto** [comment]
The default maximum-cache-age in PackageKit is G_MAXUINT, which the dnf backend correctly interprets as 'always use the cache'.

Gnome-Software seems to be manually setting the cache age for its transactions, so this doesn't interest them.
Plasma-Discover instead uses the default, which if the default is G_MAXUINT should probably be changed to something sane instead.

The question is why does the apt backend act differently? I think the dnf backend is behaving correctly here.

**Comment by aleasto** [comment]
So the apt backend ignores cache-age completely. Instead, it relies on manual RefreshCache() calls.

The dnf backend instead honours cache-age for all transactions - which is G_MAXUINT by default - including RefreshCache. That means that while for the apt backend it is enough to call `RefreshCache(false /*force*/)`, to achieve the same thing you'd need to call `RefreshCache(true /*force*/)`. The apt backend ignores the *force* parameter as without a cache-age every refresh is forced.
So fedora downstream "fixed" this issue with [this patch](https://src.fedoraproject.org/rpms/plasma-discover/blob/f38/f/discover-5.21.4-pk_refresh_force.patch), which I'm now considering adopting in upstream KDE

Overall I'm quite unhappy with how differently different backends behave...

**Comment by ximion** [comment]
> So the apt backend ignores cache-age completely. Instead, it relies on manual RefreshCache() calls.
> 
> The dnf backend instead honours cache-age for all transactions

Yes, that's just how APT is designed vs how DNF is designed. They just work differently in that regard, and PackageKit doesn't abstract this away.

**Comment by aleasto** [comment]
I think it is an issue that `RefreshCache()` - anagolously `pkcon refresh` - by default will **never** refresh the cache with the dnf backend.
Because the default cache-age is G_MAXUINT and the default force value is false.
It is implemented correctly (aka not a bug), but very unexpected if you just want to use packagekit without knowing the internals of all backends...
